### PR TITLE
fix: show file name with invalid frontmatter errors for MDX

### LINKED ISSUE #12352: `astro build` does not report which file has an invalid frontmatter for MDX files
### Astro Info

```block
Astro                    v4.16.8
Node                     v18.20.3
System                   Linux (x64)
Package Manager          npm
Output                   static
Adapter                  none
Integrations             @astrojs/mdx
                         @astrojs/sitemap
```

### If this issue only occurs in one browser, which browser is a problem?

N/A

### Describe the Bug

In build, for MDX files, a syntax error in the frontmatter does not report the file in which it failed. This works fine for Markdown and Markdoc.

```
[GenerateContentTypesError] `astro sync` command failed to generate content collection types: can not read a block mapping entry; a multiline key may not be an implicit key
  Hint:
    Check your `src/content/config.*` file for typos.
  Error reference:
    https://docs.astro.build/en/reference/errors/generate-content-types-error/
  Stack trace:
    at new AstroError (file:///home/projects/parigygoyn.github/node_modules/astro/dist/core/errors/errors.js:37:5)
    at async syncInternal (file:///home/projects/parigygoyn.github/node_modules/astro/dist/core/sync/index.js:97:7)
    at async AstroBuilder.run (file:///home/projects/parigygoyn.github/node_modules/astro/dist/core/build/index.js:195:23)
    at async build (file:///home/projects/parigygoyn.github/node_modules/astro/dist/cli/build/index.js:39:3)
    at async cli (file:///home/projects/parigygoyn.github/node_modules/astro/dist/cli/index.js:192:5)
  Caused by:
  can not read a block mapping entry; a multiline key may not be an implicit key
    at generateError (/home/projects/parigygoyn.github/node_modules/gray-matter/node_modules/js-yaml/lib/js-yaml/loader.js:167:10)
    at readBlockMapping (/home/projects/parigygoyn.github/node_modules/gray-matter/node_modules/js-yaml/lib/js-yaml/loader.js:1073:9)
    at readDocument (/home/projects/parigygoyn.github/node_modules/gray-matter/node_modules/js-yaml/lib/js-yaml/loader.js:1525:3)
    at load (/home/projects/parigygoyn.github/node_modules/gray-matter/node_modules/js-yaml/lib/js-yaml/loader.js:1614:19)
    at module.exports (/home/projects/parigygoyn.github/node_modules/gray-matter/lib/parse.js:12:17)
```

### What's the expected result?

It reports the file that failed.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-2dp4ww?file=src%2Fcontent%2Fblog%2Fusing-mdx.mdx (make sure to run `npm run build`, as dev exhibit a different behavior)

### Participation

- [X] I am willing to submit a pull request for this issue.

### PR DESCRIPTION
**Pull request by apatel369**
## Changes

- What does this change?
-`astro build` does not report which file has an invalid frontmatter for MDX files.
This PR fixes that. 

After the fix, the error looks like the below:

Screenshot 2024-11-02 at 10 15 14 AM

Closes #12352 

## Testing

It was manually tested by running a local build.

## Docs

Not needed as it just enhances existing error messages.

### CODE REVIEW

File: .changeset/wise-bulldogs-jump.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'astro': patch
+---
+
+Improves error reporting for invalid frontmatter in MDX files during the ```astro build``` process. The error message now includes the file path where the frontmatter parsing failed.
```
Comment (ematipico) [inline .changeset/wise-bulldogs-jump.md:5]:
```suggestion
Improves error reporting for invalid frontmatter in MDX files during the `astro build` command. The error message now includes the file path where the frontmatter parsing failed.
```

**Review by ematipico — APPROVED** [review]

File: packages/astro/src/core/sync/index.ts [hunk 2]
```diff
@@ -232,7 +234,7 @@ async function syncContentCollections(
 			{
 				...AstroErrorData.GenerateContentTypesError,
 				hint,
-				message: AstroErrorData.GenerateContentTypesError.message(safeError.message),
+				message: AstroErrorData.GenerateContentTypesError.message(`${safeError.message} in ${fileId}`),
```
Comment (Princesseuh) [inline packages/astro/src/core/sync/index.ts:237]: instead of using the file id in the message, you could pass it under `loc`

Comment (apatel369) [reply packages/astro/src/core/sync/index.ts]: removed fileId and added location

File: packages/astro/src/core/sync/index.ts [hunk 3]
```diff
@@ -232,7 +233,8 @@ async function syncContentCollections(
 			{
 				...AstroErrorData.GenerateContentTypesError,
 				hint,
-				message: AstroErrorData.GenerateContentTypesError.message(safeError.message),
+				message: AstroErrorData.GenerateContentTypesError.message(`${safeError.message}`),
```
Comment (bluwy) [inline packages/astro/src/core/sync/index.ts:236]: Can we revert the change on this line?

**Review by bluwy — COMMENTED** [review]
Looks like the error now shows the file name on the stack trace now: 
image

Though for some reason the error is getting printed twice, and the second error didn't have that information. But anyways the first is fixed here.

<details>
<summary>Second error</summary>

image


</details>

File: packages/astro/src/core/sync/index.ts (hunk 3 above)
Comment (apatel369) [reply packages/astro/src/core/sync/index.ts]: reverted.

**Comment by apatel369** [comment]
> Looks like the error now shows the file name on the stack trace now: image
> 
> Though for some reason the error is getting printed twice, and the second error didn't have that information. But anyways the first is fixed here.
> 
> Second error

It seems to be an existing issue with how AstroError uses location. Do I need to address this in this PR, or should I pass the location information differently?

**Comment by Princesseuh** [comment]
Pushed a few changes to improve errors with sync in general, it was using a lot of custom code to do things we already have.

- Error message is now only shown once
- Location is included in the CLI when an error crashes the process
- Reworded error message's hint to make it more clear it can be caused by an error in a file

image

File: packages/astro/src/cli/index.ts [hunk 4]
```diff
@@ -133,8 +133,8 @@ async function runCommand(cmd: string, flags: yargs.Arguments) {
 		}
 		case 'sync': {
 			const { sync } = await import('./sync/index.js');
-			const exitCode = await sync({ flags });
-			return process.exit(exitCode);
+			await sync({ flags });
+			return;
```
Comment (Princesseuh) [inline packages/astro/src/cli/index.ts:137]: We already have a global handler for CLI that handles more stuff (telemetry, error formatting etc) so this wasn't really useful.

File: packages/astro/src/core/messages.ts [hunk 5]
```diff
@@ -298,6 +298,11 @@ export function formatErrorMessage(err: ErrorWithMetadata, showFullStacktrace: b
 		output.push(`    ${cyan(underline(docsLink))}`);
 	}
 
+	if (showFullStacktrace && err.loc) {
+		output.push(`  ${bold('Location:')}`);
+		output.push(`    ${underline(`${err.loc.file}:${err.loc.line ?? 0}:${err.loc.column ?? 0}`)}`);
+	}
```
Comment (Princesseuh) [inline packages/astro/src/core/messages.ts:304]: This was something removed by Fred's previous pass on errors making them shorter 😃

I added it back, but only when the full stack trace is shown (which is either in debug in dev, or on full crashes in the CLI), so it should be shown in only contexts where it's not available elsewhere.
